src/shims/unix/macos/foreign_items.rs

```diff
@@ -166,22 +166,37 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 let [name] = this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
                 let thread = this.pthread_self()?;
                 let max_len = this.eval_libc("MAXTHREADNAMESIZE").to_target_usize(this)?;
-                let res = this.pthread_setname_np(
+                let res = if this.pthread_setname_np(
                     thread,
                     this.read_scalar(name)?,
                     max_len.try_into().unwrap(),
-                )?;
+                )? {
+                    Scalar::from_u32(0)
+                } else {
+                    this.eval_libc("ENAMETOOLONG")
+                };
                 // Contrary to the manpage, `pthread_setname_np` on macOS still
                 // returns an integer indicating success.
                 this.write_scalar(res, dest)?;
             }
             "pthread_getname_np" => {
                 let [thread, name, len] =
                     this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
-                let res = this.pthread_getname_np(
+
+                // The function's behavior isn't portable between platforms.
+                // In case of macOS, the result doesn't depend on the output
+                // bufer length. It inherits implementation from FreeBSD, but
+                // on top of it adds the thread pointer validation which can
+                // result in `ESRCH` error code.
+                //
+                // Since all other function implementations don't validate
+                // the provided thread at all, it's not happening here too.
```

Comment:
Unfortunately, it means reworking the whole thread manager and making it generation based because for FreeBSD and macOS pointers to the thread data are used, but Miri happily lives in a world of indices. Do you still want a `FIXME` instead of an issue?

Reply:
We can have both. :)

> it means reworking the whole thread manager and making it generation based because for FreeBSD and macOS pointers to the thread data are used

I don't understand what you mean by this.

Reply:
Never mind, I just checked the source of the `ThreadManager` implementation and found that in never removes threads from the vector, so a situation I had in my mind isn't possible. It was about reusing a preserved identifier of a terminated thread.